### ISSUE TITLE
What is the significance of "self.is_test" ?

### ISSUE THREAD

**Issue by tanmayGIT** [nianticlabs/wavelet-monodepth#1]
Hi,
I have well understood the flow then, I can see that :
"self.is_test" always remain False, hence the control will never enter in "If" condition
https://github.com/nianticlabs/wavelet-monodepth/blob/5bc193957056a5bab6cbc1e052f6a443279f8335/NYUv2/data.py#L133

Could you please tell me why you are doing either multiply by 1000 or dividing by 1000. 

Another point, as I am working with the smaller version of the NTU dataset i.e. "nyu_depth_v2_labeled.mat", whether this step of divide or multiply is also applicable in this dataset also ?

**Comment by MichaelRamamonjisoa** [comment]
Hi, 
Indeed, this has been inherited from DenseDepth experimental code, and is no longer needed.
The rescaling by 1000 is already discussed in detail in the DenseDepth repository issues (https://github.com/ialhashim/DenseDepth/issues/32, https://github.com/ialhashim/DenseDepth/issues/36, https://github.com/ialhashim/DenseDepth/issues/69, and a few others). Please check this first as you may find answers to a lot of your questions there.

Their train and test data were not stored with the same format (probably to reduce space requirements for the training set).
- **test data** are stored in **millimeters** using 16bit PNGs (so a division by 1000 has to be applied to get back to metric scale), 
- **train data** is quantized and stored in **0.1*255 meters** (using 8bit PNGs). The 255 factor is used during quantization to normalize the data between 0-1. The 0.1 factor is because 1 in the normalize range represents 10 meters. 

Then, as you mentioned, I'm multiplying by 1000 the depth ground truth in the dataloader, the network learns to predict depth in a scale that is 0.1 * 255 * 1000 too large. 

The division by 255 is done in the dataloader which is used here during training **only**:
https://github.com/nianticlabs/wavelet-monodepth/blob/5bc193957056a5bab6cbc1e052f6a443279f8335/NYUv2/data.py#L171

At evaluation time since the data is already in metric scale, dividing by 255 is not necessary. In order to match the original "nyu_depth_v2_labeled.mat" data (**which is in metric scale**), I need only to divide by 0.1*1000=100 to convert it back to meters: https://github.com/nianticlabs/wavelet-monodepth/blob/5bc193957056a5bab6cbc1e052f6a443279f8335/NYUv2/utils.py#L219

**Comment by tanmayGIT** [comment]
Thanks a lot for your reply. If I have understood well then if I am using  "nyu_depth_v2_labeled.mat" dataset, which I got from : https://cs.nyu.edu/~silberman/datasets/nyu_depth_v2.html. 
Here it is mentioned that the depth values are in meters 
"The values of the depth elements are in meters". I checked, the depth values are within 0-10m (the minimum value is 0.7 and the maximum value is 9.99) 
In this case, I don't need to do any sort of normalization (except putting the values within the range between 0-1)? As it is also mentioned in : https://github.com/ialhashim/DenseDepth/issues/32#issuecomment-507011482

For test data in 'nyu_data.zip', as you said that the depth values are in millimeters. Hence to get the values in meters, we need to divide by 1000. Now, we will have the values in meters. Then the range of depth values will be 0-10m? if yes then we need to normalize it within 0-1?    

For the train data in  'nyu_data.zip', as you mentioned that the values are quantized and stored where the max value is 0.1m? (i.e. the range of values is 0.00 - 0.1). Hence, if we multiply by 1000 then the maximum value becomes 100 and the range becomes 0-100m. Then to bring it back within the range of 0-1, you divided it by 100; is it?

By the way, I couldn't find the weblink to download the dataset : 'nyu_data.zip' and to see the descriptions about depth values. 
I can only get it from https://github.com/ialhashim/DenseDepth#data
but there is no description about the depth values and image type etc. 
If you know then could you please send me?
Thanks in advance !

**Comment by MichaelRamamonjisoa** [comment]
Hi,

**nyu_data.zip**
From the **data** section of [DenseDepth](https://github.com/ialhashim/DenseDepth): https://tinyurl.com/nyu-data-zip 
From the **evaluation** section: https://s3-eu-west-1.amazonaws.com/densedepth/nyu_test.zip
If you want to do an analysis on ranges, just unzip the files and load them with python as numpy arrays, for example.

**normalization**

Sorry if the DenseDepth repo got you confused, at first I got confused too. But I don't think it is relevant in this repo to understand what they have done. In short, you don't need to normalize depth between 0 and 1. DenseDepth computes losses in disparity space, and we don't. 

So: 
> Then the range of depth values will be 0-10m? 

Yes, and always.
> if yes then we need to normalize it within 0-1?

No. All we need is to have everything in the same scale when comparing things together. There are two times when we are comparing values: training, and evaluation. During training, depth is loaded from the nyu_data.zip then gets multiplied by 1000, and was originally a factor 0.1 away from metric scale, so we know that if we do nothing, the network will learn to predict depth values that will be multiplied by 0.1*1000. This means that the [0-10m] range will be represented with **values** between [0, 1000]. To get back to **meters** as in "nyu_depth_v2_labeled.mat", we divide by 100.

> For the train data in 'nyu_data.zip', as you mentioned that the values are quantized and stored where the max value is 0.1m?

No, the max depth value is still 10m, and will be **represented** by 255 (with the 8bits quantization). After quantization of the 0-10meter range with 8bits, 255 is 10meters, and 18 is 0.7meters.

> Then why to do all these unnecessary normalizations

Because I took DenseDepth's code and only did a quick adaptation. Hence my networks have been trained to predict depths that need to be divided by 100 to get back to metric depth.

**Comment by tanmayGIT** [comment]
Thanks a lot for your nice and detailed explanation. Things are now getting clear to me. 
Sorry for bothering you, I still have some queries :-(

*This comment has been edited*

What do you think I need to do if I use  "nyu_depth_v2_labeled.mat" dataset, where the values are already in meter and the range of values is 0-10m?
Should I just keep the value as it is and train the network or should I **normalize between 0-1 (by using 10 as max value and 0 as min value)** and then train ? I think both should work ? 

Hence, for this "nyu_depth_v2_labeled.mat", I think I don't need to do all these aforementioned strange normalizations etc.?
If, I will use "nyu_data.zip", then I will do the aforementioned normalization (l will just follow your way :-) )?

Another thing, which I found strange. According to your suggestion, I tried to read the original depth values from "nyu_data.zip" dataset. I get the over all min and max values in "overall_max_depth" and "overall_min_depth" variable as **255 and 18**. Hence, the original depth values are within **0-255** (not 0-100). Then what are these values ? are they quantized ? Then they are not **.1** **factor apart ?** . Do you think the code is okay or there are some issues in the code ? 
```
def readAllImagesDepth(nyu_dataset, data):
    keep_max_depth = []
    keep_min_depth = []
    for i in range(len(nyu_dataset)):
        sample = nyu_dataset[i]
        image = Image.open(BytesIO(data[sample[0]]))
        depth = Image.open(BytesIO(data[sample[1]]))

        image1 = processImage(image)
        depth1 = processImage(depth)
        keep_max_depth.append(np.max(depth1.numpy()))
        keep_min_depth.append(np.min(depth1.numpy()))

    overall_max_depth = max(keep_max_depth)
    overall_min_depth = min(keep_min_depth)
    return overall_max_depth


def processImage(pic):
    if not (_is_pil_image(pic) or _is_numpy_image(pic)):
        raise TypeError(
            'pic should be PIL Image or ndarray. Got {}'.format(type(pic)))

    if isinstance(pic, np.ndarray):
        img = torch.from_numpy(pic.transpose((2, 0, 1)))
        return img

    # handle PIL Image
    if pic.mode == 'I':
        img = torch.from_numpy(np.array(pic, np.int32, copy=False))
    elif pic.mode == 'I;16':
        img = torch.from_numpy(np.array(pic, np.int16, copy=False))
    else:
        img = torch.ByteTensor(
            torch.ByteStorage.from_buffer(pic.tobytes()))
    # PIL image mode: 1, L, P, I, F, RGB, YCbCr, RGBA, CMYK
    if pic.mode == 'YCbCr':
        n_channel = 3
    elif pic.mode == 'I;16':
        n_channel = 1
    else:
        n_channel = len(pic.mode)
    img = img.view(pic.size[1], pic.size[0], n_channel)
    img = img.transpose(0, 1).transpose(0, 2).contiguous()

    return img
```
```
def analyzeTrainingTestingData():
    data, nyu2_train = loadZipToMem('/home/tmondal/Documents/Dataset/Dehazing/nyu_data.zip')
    readAllImagesDepth(nyu2_train, data)
```
Thanks a lot for your help and cooperation.

**Comment by MichaelRamamonjisoa** [comment]
Hi,

*this answer has been edited*

>Hence to get back the values in the metric scale, we need to divide by 100, I agree!

Awesome ! :)

>In a simpler way, what do you think I need to do if I use "nyu_depth_v2_labeled.mat" dataset, where the values are already in meter and the range of values is 0-10m?
Should I just keep the value as it is and train the network or should I normalize between 0-1 (by using 10 as max value and 0 as min value) and then train ? I think both should work ?

The network will learn to produce outputs in the range you tell it to. So if you train your network using ground truth depth in metric scale, the network will also learn to produce outputs in metric scale. So you can drop all the normalizations indeed as ground truth loaded from "nyu_depth_v2_labeled.mat" is already metric depth in float. **Be careful not to divide by 255 in your loader !**

**Note that training on the reduced size nyu_depth_v2_labeled.mat dataset will not produce as good results, and has not been tested.**

**Comment by MichaelRamamonjisoa** [comment]
Hi, regarding your code, everything seems fine, it just means that the train data is ranging between 1.8 meters and 25.5 meters.

**Comment by MichaelRamamonjisoa** [comment]
Hi, 
I have realized that in their loader, the depth is first divided by 255, then multiplied by 10 meters. 

In that case, in their dataset, 255 represents 10 meters. And 18 is actually 18/255*10 which is 0.7meters. So in the above I was originally wrong, the factor between stored and metric depth is 255/10. Sorry about the confusion, l will edit the replies.

**Comment by MichaelRamamonjisoa** [comment]
Hi @tanmayGIT, I have updated my initial comment. I think a lot of the confusion (including mine) came from me not noticing that the authors were dividing by 255. If you think this is enough, do you think I can delete the other comments, as they might confuse others?

**Comment by tanmayGIT** [comment]
> Hi @tanmayGIT, I have updated my initial comment. I think a lot of the confusion (including mine) came from me not noticing that the authors were dividing by 255. If you think this is enough, do you think I can delete the other comments, as they might confuse others?

Thanks a lot for the help and time you gave for our discussion. Yes, sure. You can delete other comments.  Just keep the important comments and discussion. Which will help others also to understand it.  

Thanks for your suggestion about the quality of results which is better for "nyu_data.zip"  than "nyu_depth_v2_labeled.mat". I plan to try both for experiments. As for my case, I am not looking to propose a competitive algorithm for better depth estimation but to **use** a robust and simple technique. It will be better if I can show results on both datasets. 

I find another repository, where they use  "nyu_depth_v2_labeled.mat", and has shown how to use it. May it may help others. 

https://github.com/shirgur/UnsupervisedDepthFromFocus/blob/d41bc08a290623be8a96fa177ea769a7ef742dc2/Utils/Datasets.py#L66

**Comment by tanmayGIT** [comment]
> Hi,
> 
> I'll try to answer quickly as I thing most of your questions are related to data, which I didn't produce. You should probably ask the authors of DenseDepth all of these queries.
> 
> I think you are misunderstanding what quantization means. Here it simply means rounding to the nearest integer, so that you can store with an integer value between 0 and 255, which is 8bits.
> 
> > as an example depth value of 7.89 is first quantized so it becomes 78.9 and then stored (as .1 factor away) as 7.89 (same as multiplying 78.9 by .1 or dividing it by 10 ) ? I assume that these are the values (i.e 7.89) stored in "nyu_data.zip" ?
> 
> No, 7.89 meters are **represented** with the value 79. Because it's multiplied by 10, then rounded to the nearest integer, or **quantized**.
> 
> > Hence to get back the values in the metric scale, we need to divide by 100, I agree!
> 
> Awesome ! :)
> 
> > But the obtained values will be still quantized value (not the original values, because we did nothing to reconvert from quantized value to normal value) ?
> 
> Refer to the first remark, quantization is not just about multiplying by a factor.
> 
> > In a simpler way, what do you think I need to do if I use "nyu_depth_v2_labeled.mat" dataset, where the values are already in meter and the range of values is 0-10m?
> > Should I just keep the value as it is and train the network or should I normalize between 0-1 (by using 10 as max value and 0 as min value) and then train ? I think both should work ?
> 
> The network will learn to produce outputs in the range you tell it to. So if you train your network using ground truth depth in metric scale, the network will also learn to produce outputs in metric scale. So you can drop all the normalizations indeed.
> 
> **Note that training on the reduced size nyu_depth_v2_labeled.mat dataset will not produce as good results, and has not been tested**
> 
> For the rest of your queries, please ask the creators of the dataset. Let me know if you have other queries not related to data

**Comment by MichaelRamamonjisoa** [comment]
Thanks for your suggestion! 
I have now updated all comments. If everything looks clear now, you can close this issue. Sorry again for the misunderstanding, I will put a small note about the normalization process that is happening in the readme

**Comment by MichaelRamamonjisoa** [comment]
Left a note in [NYUv2/README.md](https://github.com/nianticlabs/wavelet-monodepth/tree/main/NYUv2/README.md) (commit https://github.com/nianticlabs/wavelet-monodepth/commit/57a4602b9591bd048593f13cab4c8e300cb61f62), closing.

**Comment by tanmayGIT** [comment]
> Hi,
> I have realized that in their loader, the depth is first divided by 255, then multiplied by 10 meters.
> 
> In that case, in their dataset, 255 represents 10 meters. And 18 is actually 18/255*10 which is 0.7meters. So in the above I was originally wrong, the factor between stored and metric depth is 255/10. Sorry about the confusion, l will edit the replies.

Hi Again, 
I was out of this work for some time.. Now, I am back on this topic. Sorry to reinitiate the discussion again. As you mentioned, the original depth values are between 0-10m. They have transformed it within the range of 0-255 and created the dataset. 
You made small typos: it will be (18*10)/255 = 0.7 m. 
In dataloader, we first divide it by 255 hence the values become within the range of 0-1 then it is clamped between 10-1000 or between 0.1m-10m. 
Hence, the sequence is the original values: [0m-10m] -> [0-255] -> [0-1](dividing by 255) -> [0-1000](multiplying by 1000) -> [0m-10m](in metric scale) -> [10-1000](by using torch clamp); ref : https://github.com/nianticlabs/wavelet-monodepth/blob/81164835bea493981d557d60c826ac35415a8cf2/NYUv2/data.py#L138

now to compute the loss of predicted depth (which is in range 0-1), you are dividing (1000/100) by depth values (which are within the range: 10-1000). ref: https://github.com/nianticlabs/wavelet-monodepth/blob/81164835bea493981d557d60c826ac35415a8cf2/NYUv2/utils.py#L24
in DenseDepth, they divide 1000 by depth values; ref: https://github.com/ialhashim/DenseDepth/blob/af32304895e17a35fcda87182b4459e77684f749/PyTorch/utils.py#L5 
Could you please tell, why you did like this and why they did like this (your opinion)?

Another point, in densedepth, they normalize the predicted image (I think it is done for visualization purpose only); ref  https://github.com/ialhashim/DenseDepth/blob/af32304895e17a35fcda87182b4459e77684f749/PyTorch/train.py#L110
But, you did it on the GT depth image, maybe it was needed for your case, I suppose ?; ref: https://github.com/nianticlabs/wavelet-monodepth/blob/81164835bea493981d557d60c826ac35415a8cf2/NYUv2/train.py#L63
